File: src/rust/multiplier.rs

struct Entry {
    first: u64,
    second: u64,
}

struct PeasantTables {
    next: Entry,
}

impl Iterator for PeasantTables {
    type Item = Entry;
    fn next(&mut self) -> Option<Entry> {
        let current = Entry { ..self.next };

        self.next = Entry {
            first: current.first.saturating_mul(2),
            second: current.second / 2,
        };

        match current.second {
            0 => None,
            _ => Some(current),
        }
    }
}

pub fn multiply(a: u64, b: u64) -> Option<u64> {
    let tables = PeasantTables {
        next: Entry {
            first: a,
            second: b,
        },
    };

    return tables.filter(|e| e.second % 2 != 0)
        .map(|e| e.first)
        .fold(Some(0), |a, c| {
            println!("{:?}, {}, {}", a, c, std::u64::MAX);

            if a.is_none() || c == std::u64::MAX {
                None
            } else {
                a.unwrap().checked_add(c)
            }
        });
}

#[cfg(test)]
mod tests {
    use super::multiply;
    use std::u64::MAX;

    #[test]
    fn test_multiplier() {
        assert_eq!(multiply(3, 2), Some(6));
        assert_eq!(multiply(3, 4), Some(12));
        assert_eq!(multiply(3, 0), Some(0));
        assert_eq!(multiply(0, 44), Some(0));
        assert_eq!(multiply(300, 5), Some(1500));
        assert_eq!(multiply(MAX, 2), None)
    }
}

File: src/swift/multiplier.swift
struct Entry {
    let first: UInt64
    let second: UInt64
}

class PeasantTables: Sequence, IteratorProtocol {
    var current: Entry

    init(seed: Entry) {
        current = seed
    }

    func next() -> Entry? {
        guard current.second > 0 else { return nil }
        defer { current = Entry(first: current.first * 2, second: current.second / 2) }
        return current
    }
}

func multiply(_ one: UInt64, _ another: UInt64) -> UInt64 {
    let skipEvenTables: (Entry) -> Bool = { $0.second % 2 != 0 }
    let collectFirst: (Entry) -> UInt64 = { $0.first }

    let tables = PeasantTables(seed: Entry(first: one, second: another))
    return tables.filter(skipEvenTables).map(collectFirst).reduce(0, +)
}
